include/seastar/core/internal/io_request.hh

```diff
@@ -422,18 +422,23 @@ struct io_request::part {
 
 // Helper pair of IO direction and length
 struct io_direction_and_length {
-    size_t _directed_length; // bit 0 is R/W flag
+    size_t _directed_length; // bits 0 and 1 depict R/W/D flags
+
+    static constexpr int direction_bits_count = 2;
+    static constexpr int direction_bits_mask = (1 << direction_bits_count) - 1;
 
 public:
-    size_t length() const noexcept { return _directed_length >> 1; }
-    int rw_idx() const noexcept { return _directed_length & 0x1; }
+    size_t length() const noexcept { return _directed_length >> direction_bits_count; }
+    int rwd_idx() const noexcept { return _directed_length & direction_bits_mask; }
+
+    static constexpr int discard_idx = 2;
```

Comment:
I think that discard request should carry write_idx direction and whether or not to submit it via sink or via punch_hole() should be decided based on request::op()

Reply:
Sure, I can revert the change and use `io_direction_and_length::write` for discards. The new direction does not look right in the case of `duplex` mode -- I would not expect three directions in duplex mode.

However, if there is no `discard` direction, I see the following three issues that are described below.

Firstly, if the write direction is used, then the calculation in `internal::request_tokens()` will need to contain a new if statement. It is going to distinguish between true write requests and discard requests (based on operation type), because the coefficients are different for write/discard operations.

Secondly, I am not sure about the usage of `_group->_max_request_length[]`. If I understand correctly, the maximum request length is different for discards and for writes. So, simple usage of `_group->_max_request_length[direction]` will be invalid. Somehow, I feel that it would end up with creation of a new private function that would look similar to the function below:

```
size_t get_max_request_length(io_direction_and_length dnl, const internal::io_request& req) {
    if (req.opcode() != internal::io_request::operation::discard) {
    	return _group->_max_request_length[dnl.rw_idx()];
    }
    
    return _group->_max_discard_request_length;
}
```

Lastly, the way how statistics are accumulated in `io_queue::priority_class_data::on_dispatch()` will mix discards and writes. The code looks as follows: `_rwstat[dnl.rwd_idx()].add(dnl.length());`.

If the three mentioned points do not cause a problem, then I will introduce the change.